diff --git forkSrcPrefix/worker-core/wasm_exec.js forkDstPrefix/worker-core/wasm_exec.js
index defc73ba8201fdd3452adcddddeb8cd8d6131599..55ae0a1e9d96c1c9a2590d5f4280ca154e2bbbd2 100644
--- forkSrcPrefix/worker-core/wasm_exec.js
+++ forkDstPrefix/worker-core/wasm_exec.js
@@ -303,9 +303,25 @@
 
 					// func finalizeRef(v ref)
 					"syscall/js.finalizeRef": (v_ref) => {
-						// Note: TinyGo does not support finalizers so this should never be
-						// called.
-						console.error('syscall/js.finalizeRef not implemented');
+            // Note: TinyGo does not support finalizers so this is only called
+            // for one specific case, by js.go:jsString. and can/might leak memory.
+            // This is a workaround for the lack of finalizers in TinyGo.
+            // https://github.com/tinygo-org/tinygo/issues/1140#issuecomment-2529705551
+
+            // magicValue taken from `unboxValue` function.
+            const magicValue = 0xffffffffn;
+            const id = v_ref & magicValue;
+            if (this._goRefCounts?.[id] !== undefined) {
+              this._goRefCounts[id]--;
+              if (this._goRefCounts[id] === 0) {
+                const v = this._values[id];
+                this._values[id] = null;
+                this._ids.delete(v);
+                this._idPool.push(id);
+              }
+            } else {
+              console.log("syscall/js.finalizeRef: unknown id", id);
+            }
 					},
 
 					// func stringVal(value string) ref
